### ISSUE TITLE
Add with_lora_target_modules() method for advanced LoRA configuration

### ISSUE THREAD

**Issue by msalganik** [msalganik/torchtune_config_writer#1]
## Summary

Add a convenience method for configuring which layers LoRA is applied to, including attention modules and MLP layers.

## Motivation

Research shows that **LoRA performs better when applied to all weight matrices, especially MLP and MoE layers** (see references below). Currently, users need to use `override()` to configure target modules:

```python
# Current workaround
builder.override({
    "model": {
        "lora_attn_modules": ["q_proj", "k_proj", "v_proj", "output_proj"],
        "apply_lora_to_mlp": True,
        "apply_lora_to_output": False
    }
})
```

## Proposed API

```python
def with_lora_target_modules(
    self,
    attn_modules: List[str] = None,
    apply_to_mlp: bool = None,
    apply_to_output: bool = None
) -> 'TorchtuneConfigBuilder':
    """
    Configure which layers LoRA is applied to.
    
    Args:
        attn_modules: Which attention projections to apply LoRA to.
                     Options: ["q_proj", "k_proj", "v_proj", "output_proj"]
                     Default: Use base config's default (typically ["q_proj", "v_proj"])
        apply_to_mlp: Whether to apply LoRA to MLP layers
                     Default: Use base config's default
        apply_to_output: Whether to apply LoRA to final output projection
                        Default: Use base config's default
    
    Example:
        >>> # Apply to all attention layers + MLP (best performance)
        >>> builder.with_lora_target_modules(
        ...     attn_modules=["q_proj", "k_proj", "v_proj", "output_proj"],
        ...     apply_to_mlp=True,
        ...     apply_to_output=False
        ... )
    """
```

## Usage Examples

```python
# Example 1: Maximum coverage (research recommendation)
builder = TorchtuneConfigBuilder("llama3_1/8B_lora_single_device")
builder.with_lora_target_modules(
    attn_modules=["q_proj", "k_proj", "v_proj", "output_proj"],
    apply_to_mlp=True,
    apply_to_output=False
)
builder.with_lora_params(32, 64)  # Higher rank for more parameters

# Example 2: Attention-only (minimal approach)
builder.with_lora_target_modules(
    attn_modules=["q_proj", "v_proj"],
    apply_to_mlp=False
)

# Example 3: Only MLP layers
builder.with_lora_target_modules(
    attn_modules=[],
    apply_to_mlp=True
)
```

### Research
- **Blog Post**: [Fine-tuning with LoRA: Best Practices](https://thinkingmachines.ai/blog/lora/)
  - Key finding: "LoRA performs better when applied to all weight matrices, especially MLP and MoE layers"
  - Attention-only configurations significantly underperform even with matched parameter counts

### Torchtune Documentation
- [Fine-Tuning Llama2 with LoRA Tutorial](https://docs.pytorch.org/torchtune/stable/tutorials/lora_finetune.html)
  - Documents `lora_attn_modules` parameter
  - Default: `["q_proj", "v_proj"]` with rank 8
  - Recommendation: "some experiments have found it beneficial to apply LoRA to all linear layers in self-attention and increase rank to 16 or 32"

### Config Examples
From `llama3_1/8B_lora_single_device.yaml`:
```yaml
model:
  _component_: torchtune.models.llama3_1.lora_llama3_1_8b
  lora_attn_modules: ['q_proj', 'v_proj', 'output_proj']
  apply_lora_to_mlp: True
  apply_lora_to_output: False
  lora_rank: 8
  lora_alpha: 16
```

## Implementation Notes

**Category**: Scientific parameter (affects model architecture and capacity)

**Config paths modified**:
- `model.lora_attn_modules` (list of strings)
- `model.apply_lora_to_mlp` (boolean)
- `model.apply_lora_to_output` (boolean)

**Validation**:
- `attn_modules` must be list of valid projection names
- Valid options: `{"q_proj", "k_proj", "v_proj", "output_proj"}`
- Can be empty list (if using MLP-only)

**Default behavior**:
- If parameter is `None`, don't modify that field (preserve base config default)
- This allows selective override

## Priority

**Phase 2+** - This is an advanced configuration option. Most users will be fine with base config defaults. Add when usage patterns show demand for this level of control.

## Related Issues

This relates to the broader question of providing convenience methods vs. encouraging `override()` for advanced use cases. See design rationale in SPEC.md and Appendix C.

---

**Labels**: enhancement, phase-2, scientific-parameter
